File: daq/slc/extra/javalib/b2daq/src/b2daq/dqm/core/DoubleArray.java
package b2daq.dqm.core;

import java.io.IOException;

import b2daq.core.Reader;
import b2daq.core.Writer;

public class DoubleArray extends NumberArray {

	private double[] _value_v;
	
	public DoubleArray() {
		_value_v = new double [0];
		for (int n = 0; n < length(); n++ ) _value_v[n] = 0;
	}

	public DoubleArray(int len) {
		_value_v = new double [len];
		for (int n = 0; n < length(); n++ ) _value_v[n] = 0;
	}

	public DoubleArray(NumberArray array) { copy(array); }

	public void copy(NumberArray array) {
		_value_v = new double [array.length()];
		for (int n = 0; n < length(); n++ ) _value_v[n] = array.get(n);
	}

	public void copy(double [] array) {
		_value_v = new double [array.length];
		for (int n = 0; n < length(); n++ ) _value_v[n] = array[n];
	}

	public double get(int index) {
		if ( index >= 0 && index < length() ) return _value_v[index];
		else return -1;
	}

	public void set(int index, double data) {
		if ( index >= 0 && index < length() ) _value_v[index] = data;
	}

	public int length() { return _value_v.length; }

	public void resize(int len) {
		_value_v = new double [len];
		for (int n = 0; n < length(); n++ ) _value_v[n] = 0;
	}

	public void writeSingle(double value, Writer writer) throws IOException {
		writer.writeDouble(value);
	}

	public double readSingle(Reader reader) throws IOException {
		return reader.readDouble();
	}

}

File: daq/slc/dqm/src/DoubleArray.cc
#include "daq/slc/dqm/DoubleArray.h"

#include <sstream>

using namespace Belle2;

DoubleArray::DoubleArray() throw()
  : NumberArray(0), m_value_v(NULL) {}

DoubleArray::DoubleArray(size_t size) throw()
  : NumberArray(), m_value_v(NULL)
{
  resize(size);
}

DoubleArray::DoubleArray(const NumberArray& v) throw()
  : NumberArray(), m_value_v(NULL)
{
  resize(v.size());
  for (size_t n = 0; n < size(); n++) set(n, m_value_v[n]);
}

DoubleArray::~DoubleArray() throw()
{
  if (m_value_v != NULL) delete [] m_value_v;
}

double DoubleArray::get(size_t n) const throw()
{
  if (n < size()) return m_value_v[n];
  else return -1;
}
void DoubleArray::set(size_t n, double v) throw()
{
  if (n < size()) m_value_v[n] = (double)v;
}
void DoubleArray::resize(int s) throw()
{
  if (m_value_v != NULL) delete [] m_value_v;
  m_value_v = new double [s];
  m_size = s;
  for (size_t n = 0; n < size(); n++) set(n, 0);
}

std::string DoubleArray::toScript() const throw()
{
  std::stringstream ss("");
  for (size_t n = 0; n < size(); n++) {
    ss << get(n);
    if (n < size() - 1) ss << ",";
  }
  return ss.str();
}

double& DoubleArray::operator[](size_t n) throw()
{
  if (n < size()) return m_value_v[n];
  return m_value_v[0];
}

double DoubleArray::operator[](size_t n) const throw()
{
  if (n < size()) return m_value_v[n];
  return 0;
}

void DoubleArray::readObject(Reader& reader)
throw(IOException)
{
  for (size_t n = 0; n < size(); n++) {
    m_value_v[n] = reader.readDouble();
  }
}

void DoubleArray::writeObject(Writer& writer)
const throw(IOException)
{
  for (size_t n = 0; n < size(); n++) {
    writer.writeDouble(m_value_v[n]);
  }
}

double DoubleArray::readElement(Reader& reader)
throw(IOException)
{
  return reader.readDouble();
}

void DoubleArray::writeElement(double value, Writer& writer)
const throw(IOException)
{
  writer.writeDouble(value);
}


